### PR TITLE
Implement LWG-4083 `views::as_rvalue` should reject non-input ranges

### LINKED ISSUE #4762: LWG-4083 `views::as_rvalue` should reject non-input ranges
LWG-4083 `views::as_rvalue` should reject non-input ranges

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #4762.

Reported a related EDG bug: DevCom-10698021.

### CODE REVIEW

File: tests/std/tests/P2446R2_views_as_rvalue/test.cpp [hunk 1]
```diff
@@ -401,6 +401,17 @@ void test_example_from_p2446r2() {
...
+// LWG-4083 "views::as_rvalue should reject non-input ranges"
+struct OutputRvalueIterator {
+    using difference_type = int;
+    int operator*() const;
+    OutputRvalueIterator& operator++();
+    void operator++(int);
+};
+using OutputRvalueRange = decltype(ranges::subrange{OutputRvalueIterator{}, unreachable_sentinel});
+
+static_assert(!CanViewAsRvalue<OutputRvalueRange>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P2446R2_views_as_rvalue/test.cpp:413]: There are EDG test failures:

```
"D:\build\out\inc\ranges", line 1527: error: template constraint not satisfied
          concept _Can_as_rvalue = requires(_Rng&& __r) { as_rvalue_view{static_cast<_Rng&&>(__r)}; };
                                                          ^
"D:\build\out\inc\utility", line 187: note: atomic constraint not constant
  concept _Different_from = !same_as<remove_cvref_t<_Ty1>, remove_cvref_t<_Ty2>>;
                            ^
          detected during:
            instantiation of
                      "std::_Choice_t<std::ranges::views::_As_rvalue_fn::_St>
                      std::ranges::views::_As_rvalue_fn::_Choose<_Rng>()
                      noexcept [with
                      _Rng=std::ranges::subrange<OutputRvalueIterator,
                      std::unreachable_sentinel_t,
                      std::ranges::subrange_kind::unsized>]" at line 1545
            instantiation of "const
                      std::_Choice_t<std::ranges::views::_As_rvalue_fn::_St>
                      std::ranges::views::_As_rvalue_fn::_Choice [with
                      _Rng=std::ranges::subrange<OutputRvalueIterator,
                      std::unreachable_sentinel_t,
                      std::ranges::subrange_kind::unsized>]" at line 413 of
                      "D:\a\_work\1\s\tests\std\tests\P2446R2_views_as_rvalue\t
                      est.cpp"
```

As a reminder, when changing a test, you should always run that test at a minimum.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: tests/std/tests/P2446R2_views_as_rvalue/test.cpp (hunk 1 above)
Comment (frederick-vs-ja) [reply tests/std/tests/P2446R2_views_as_rvalue/test.cpp:413]: Oops... I found that EDG accepts this if we switch to use direct-initialization.
```diff
diff --git a/stl/inc/ranges b/stl/inc/ranges
index 6f7e4d06..094aa1d5 100644
--- a/stl/inc/ranges
+++ b/stl/inc/ranges
@@ -1524,7 +1524,7 @@ namespace ranges {
 
     namespace views {
         template <class _Rng>
-        concept _Can_as_rvalue = requires(_Rng&& __r) { as_rvalue_view{static_cast<_Rng&&>(__r)}; };
+        concept _Can_as_rvalue = requires(_Rng&& __r) { as_rvalue_view(static_cast<_Rng&&>(__r)); };
 
         class _As_rvalue_fn : public _Pipe::_Base<_As_rvalue_fn> {
         private:
```

Comment (frederick-vs-ja) [reply tests/std/tests/P2446R2_views_as_rvalue/test.cpp:413]: Reduced and reported DevCom-10698021.

File: stl/inc/ranges [hunk 2]
```diff
@@ -1524,15 +1524,19 @@ namespace ranges {
 
     namespace views {
         template <class _Rng>
+#ifdef __EDG__ // TRANSITION, DevCom-10698021
+        concept _Can_as_rvalue = requires(_Rng&& __r) { as_rvalue_view(static_cast<_Rng&&>(__r)); };
+#else // ^^^ workaround / no workaround vvv
```
Comment (frederick-vs-ja) [inline stl/inc/ranges:1529]: Not sure whether "workaround" is appropriate here as the standard wording does say direct-initialization ([[range.as.rvalue.overview]/2.2](https://eel.is/c++draft/range.as.rvalue.overview#2.2)).

Comment (StephanTLavavej) [reply stl/inc/ranges:1529]: Seems like we should perma-workaround this by Doing What The Standard Says<sup>TM</sup>.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks - I really appreciate how you're always implementing LWG issues and reporting compiler bugs! :heart_eyes_cat: :white_check_mark: :lady_beetle:

**Pull request merged**